qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py

```diff
@@ -0,0 +1,120 @@
+    invalid_labels = [label for label, _ in vibrational_labels if
+                      not valid_vibr_label_pattern.match(label)]
+    if invalid_labels:
+        raise ValueError(f"Invalid labels: {invalid_labels}")
+
+
+def _validate_indices(vibrational_labels: List[Tuple[str, float]], num_modes: int,
+                      num_modals: List[int]):
+    for labels, _ in vibrational_labels:
+        coeff_labels_split = labels.split(" ")
+        check_list = [0] * num_modes
+        last_op, last_mode_index, last_modal_index = "+", num_modes, max(num_modals)
+        for label in coeff_labels_split:
+            op, mode_index_str, modal_index_str = re.split('[*_]', label)
+            mode_index = int(mode_index_str)
+            modal_index = int(modal_index_str)
+            if _is_index_out_of_range(mode_index, num_modes, modal_index, num_modals):
+                raise ValueError(f"Indices out of the declared range for label {label}.")
+            if _is_label_duplicated(mode_index, last_mode_index, modal_index, last_modal_index, op,
+                                    last_op):
+                raise ValueError(f"Operators in a label duplicated for label {label}.")
+            if _is_order_incorrect(mode_index, last_mode_index, modal_index, last_modal_index, op,
+                                   last_op):
+                raise ValueError(
+                    f"Incorrect order of operators for label {label} and previous label "
+                    f"{str(last_op) + '_' + str(last_mode_index) + '*' + str(last_modal_index)}.")
+
+            last_op, last_mode_index, last_modal_index = op, mode_index, modal_index
+
+            increment = 1 if op == "+" else -1
+            check_list[int(mode_index)] += increment
+        if not all(v == 0 for v in check_list):
+            raise ValueError(
+                f"Modes of raising and lowering operators do not agree for labels {labels}.")
+
+
+def _is_index_out_of_range(mode_index: int, num_modes: int, modal_index: int,
+                           num_modals: List[int]) -> bool:
+    return mode_index >= num_modes or modal_index >= num_modals[int(mode_index)]
+
+
+def _is_label_duplicated(mode_index: int, last_mode_index: int, modal_index: int,
+                         last_modal_index: int, op: str, last_op: str) -> bool:
+    return modal_index == last_modal_index and mode_index == last_mode_index and op == last_op
+
+
+def _is_order_incorrect(mode_index: int, last_mode_index: int, modal_index: int,
+                        last_modal_index: int, op: str, last_op: str) -> bool:
+    return _is_mode_order_incorrect(mode_index, last_mode_index) or \
+           _is_modal_order_incorrect(last_mode_index, mode_index, last_modal_index, modal_index) \
+           or _is_operator_order_incorrect(mode_index, last_mode_index, modal_index,
+                                           last_modal_index, op, last_op)
```

Comment:
I think you can inline those methods with a comment explaining each rather than splitting out into extra functions. But here I don't mind.